lib/Sema/ConstraintSystem.cpp

```diff
@@ -3099,6 +3031,112 @@ bool ConstraintSystem::diagnoseAmbiguityWithFixes(
   return diagnosed;
 }
 
+bool ConstraintSystem::diagnoseAmbiguityWithFixes(
+    SmallVectorImpl<Solution> &solutions) {
+  if (solutions.empty())
+    return false;
+
+  SolutionDiff solutionDiff(solutions);
+
+  if (diagnoseConflictingGenericArguments(*this, solutionDiff, solutions))
+    return true;
+
+  if (auto bestScore = solverState->BestScore) {
+    solutions.erase(llvm::remove_if(solutions,
+                                    [&](const Solution &solution) {
+                                      return solution.getFixedScore() >
+                                             *bestScore;
+                                    }),
+                    solutions.end());
+
+    if (llvm::all_of(solutions, [&](const Solution &solution) {
+          auto score = solution.getFixedScore();
+          return score.Data[SK_Fix] == 0 && solution.Fixes.empty();
+        }))
+      return false;
+  }
+
+  if (solutions.size() < 2)
+    return false;
+
+  if (diagnoseAmbiguityWithEphemeralPointers(*this, solutions))
+    return true;
+
+
+  // Algorithm is as follows:
+  //
+  // a. Aggregate all of the available fixes based on callee locator;
+  // b. For each ambiguous overload match aggregated fixes and diagnose;
+  // c. Discard all of the fixes which have been already considered
+  //    as part of overload diagnostics;
+  // d. Diagnose remaining (uniqued based on kind + locator) fixes
+  //    iff they appear in all of the solutions.
+
+  using Fix = std::pair<const Solution *, const ConstraintFix *>;
+
+  llvm::SmallSetVector<Fix, 4> fixes;
+  for (auto &solution : solutions) {
+    for (auto *fix : solution.Fixes)
+      fixes.insert({&solution, fix});
+  }
+
+  llvm::MapVector<ConstraintLocator *, SmallVector<Fix, 4>> fixesByCallee;
+
+  for (const auto &entry : fixes) {
+    const auto &solution = *entry.first;
+    const auto *fix = entry.second;
+    auto *calleeLocator = solution.getCalleeLocator(fix->getLocator());
+
+    fixesByCallee[calleeLocator].push_back({&solution, fix});
+  }
+
+  bool diagnosed = false;
+
+  // All of the fixes which have been considered already.
+  llvm::SmallSetVector<Fix, 4> consideredFixes;
+
+  for (const auto &ambiguity : solutionDiff.overloads) {
+    auto fixes = fixesByCallee.find(ambiguity.locator);
+    if (fixes == fixesByCallee.end())
+      continue;
+
+    auto aggregate = fixes->second;
+    diagnosed |= ::diagnoseAmbiguity(*this, ambiguity, aggregate, solutions);
+
+    consideredFixes.insert(aggregate.begin(), aggregate.end());
```

Comment:
I wonder if In the cases where `diagnoseAmbiguity` returns false(if there's any) ... since we are adding them independently it maybe is a problem since they would be dropped from `fixes` undiagnosed? If there are such cases, maybe we should let them to attempt diagnose by kind?

Reply:
The reason why I did that is related to the fact that it could be a bunch of different fixes all related to the same callee e.g. `1 + ""` would produce one fix for first argument and one for second. Even if we didn't produce ambiguity related to overload in such case keeping them around wouldn't help to diagnose anything later because there is going to be no intersection so it's better to produce "I don't know" diagnostic and figure out case by case what went wrong...

Reply:
Ah makes sense, I think I actually have seen situations like that too in SR-12689. I was just worried that it may end up discarding useful fixes in some way... but now I see the point :)